### PR TITLE
changed record internals from hashtable to array

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: tests/flow/test_index_updates.py [hunk 1]
```diff
@@ -27,12 +28,17 @@ def setUpClass(cls):
...
+        # cls.r = redis.Redis()
+        # redis_graph = Graph("index_test", cls.r)
+
         cls.populate_graph()
         cls.build_indices()
 
     @classmethod
     def tearDownClass(cls):
         cls.r.stop()
+        # pass
```
Comment (jeffreylovitz) [inline tests/flow/test_index_updates.py]: Why the changes in this file?

File: src/parser/ast.c [hunk 2]
```diff
@@ -62,33 +65,33 @@ AST_Validation _AST_ValidateReferredFunctions(TrieMap *referred_functions,
...
+
+  void *val;
+  char *ptr;
+  tm_len_t len;
+  TrieMapIterator *it = TrieMap_Iterate(referredEntities, "", 0);
+  // Scan through aliases and give each one an ID.
+  while(TrieMapIterator_Next(it, &ptr, &len, &val)) {
+    uint *entityID = malloc(sizeof(uint));
+    *entityID = id++;
+    TrieMap_Add(ast->_aliasIDMapping, ptr, len, entityID, TrieMap_NOP_REPLACE);
```
Comment (jeffreylovitz) [inline src/parser/ast.c]: I like this approach!

File: src/execution_plan/record.c [hunk 3]
```diff
@@ -6,56 +6,63 @@
...
-        printf("alias %s:\n", e->alias);
+GraphEntity *Record_GetGraphEntity(const Record r, int idx) {
+    Entry e = r[idx];
+    switch(e.type) {
+        case REC_TYPE_NODE:
+            return (GraphEntity*)Record_GetNode(r, idx);
+        case REC_TYPE_EDGE:
+            return (GraphEntity*)Record_GetEdge(r, idx);
+        case REC_TYPE_SCALAR: 
+            return (GraphEntity*)(Record_GetScalar(r, idx).ptrval);
```
Comment (jeffreylovitz) [inline src/execution_plan/record.c]: What do you think of asserting that the SIType of the value is a `T_PTR` here?

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 4]
```diff
@@ -37,21 +41,17 @@ OpBase *NewNodeByLabelScanOp(GraphContext *gc, Node *node) {
...
-    }
-
     bool depleted = false;
     GxB_MatrixTupleIter_next(op->iter, NULL, &nodeId, &depleted);
     if(depleted) return OP_DEPLETED;
-
-    Graph_GetNode(op->g, nodeId, op->node);
+    
+    Node *n = Record_GetNode(r, op->node_rec_idx);
+    Graph_GetNode(op->g, nodeId, n);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_node_by_label_scan.c]: Maybe add a comment here to indicate that `Graph_GetNode` is writing into the Record.

File: src/execution_plan/ops/op_index_scan.c [hunk 5]
```diff
@@ -27,14 +30,14 @@ OpBase *NewIndexScanOp(Graph *g, Node *node, IndexIter *iter) {
...
+OpResult IndexScanConsume(OpBase *opBase, Record r) {
   IndexScan *op = (IndexScan*)opBase;
 
   EntityID *nodeId = IndexIter_Next(op->iter);
   if (!nodeId) return OP_DEPLETED;
 
-  Graph_GetNode(op->g, *nodeId, op->node);
-  Record_AddEntry(r, op->node->alias, SI_PtrVal(op->node));
+  Node *n = Record_GetNode(r, op->nodeRecIdx);
+  Graph_GetNode(op->g, *nodeId, n);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_index_scan.c]: Maybe add a comment here to indicate that `Graph_GetNode` is writing into the Record.

File: src/execution_plan/ops/op_delete.c [hunk 6]
```diff
@@ -10,55 +10,21 @@
...
-
 void _LocateEntities(OpDelete *op, QueryGraph *qg, AST_DeleteNode *ast_delete_node) {
+    AST *ast = AST_GetFromLTS();
     for(int i = 0; i < Vector_Size(ast_delete_node->graphEntities); i++) {
         char *entity_alias;
         Vector_Get(ast_delete_node->graphEntities, i, &entity_alias);
 
         // Current entity is a node.
+        int entityRecIdx = AST_GetAliasID(ast, entity_alias);
         Node *n = QueryGraph_GetNodeByAlias(qg, entity_alias);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_delete.c]: I'd like to introduce a faster type-checking mechanism than this at some point.

File: src/execution_plan/ops/op_create.c [hunk 7]
```diff
@@ -280,20 +285,11 @@ OpResult OpCreateConsume(OpBase *opBase, Record *r) {
     }
 
     if(op->op.childCount == 0) return OP_DEPLETED;
-    else return res;
+    return res;
 }
 
 OpResult OpCreateReset(OpBase *ctx) {
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_create.c]: Is it safe for this reset to be a no-op?

File: src/execution_plan/ops/op_create.c [hunk 8]
```diff
@@ -87,40 +89,40 @@ OpBase* NewCreateOp(RedisModuleCtx *ctx, GraphContext *gc, AST_Query *ast, Query
...
         /* Create a new node. */
         Node *newNode = Node_New(n->label, n->alias);
 
         /* Save node for later insertion. */
-        Vector_Push(op->created_nodes, newNode);
+        op->created_nodes = array_append(op->created_nodes, newNode);
 
         /* Update record with new node. */
-        Record_AddEntry(r, newNode->alias, SI_PtrVal(newNode));
+        Record_AddScalar(r, op->nodes_to_create[i].node_rec_idx, SI_PtrVal(newNode));
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_create.c]: Why aren't we using Record_AddNode and `*newNode`?

File: src/execution_plan/ops/op_conditional_traverse.c [hunk 9]
```diff
@@ -121,34 +116,34 @@ OpResult CondTraverseConsume(OpBase *opBase, Record *r) {
...
+            srcNode = Record_GetNode(r, op->srcNodeRecIdx);
+            destNode = Record_GetNode(r, op->destNodeRecIdx);
         }
 
         Graph_GetEdgesConnectingNodes(op->graph,
                                       ENTITY_GET_ID(srcNode),
                                       ENTITY_GET_ID(destNode),
                                       op->edgeRelationType,
                                       &op->edges);
-        return _CondTraverse_SetEdge(op, r);
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_conditional_traverse.c]: What if `_CondTraverse_SetEdge` returns OP_DEPLETED?

File: src/execution_plan/ops/op_cond_var_len_traverse.c [hunk 10]
```diff
@@ -71,7 +72,7 @@ OpResult CondVarLenTraverseConsume(OpBase *opBase, Record *r) {
 
     // op->ae->dest_node is already in record
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_cond_var_len_traverse.c]: Outdated comment

File: src/arithmetic/arithmetic_expression.h [hunk 11]
```diff
@@ -99,6 +99,7 @@ typedef struct {
 		SIValue constant;
 		struct {
 			char *entity_alias;
+            int entity_alias_idx;
```
Comment (jeffreylovitz) [inline src/arithmetic/arithmetic_expression.h]: Should be tabs

File: tests/flow/test_index_updates.py (hunk 1 above)
Comment (swilly22) [reply tests/flow/test_index_updates.py]: Adding the ability to run this test against a none disposable redis-server

File: src/execution_plan/record.c (hunk 3 above)
Comment (swilly22) [reply src/execution_plan/record.c]: 👍

File: src/execution_plan/ops/op_node_by_label_scan.c (hunk 4 above)
Comment (swilly22) [reply src/execution_plan/ops/op_node_by_label_scan.c]: 👍

File: src/execution_plan/ops/op_index_scan.c (hunk 5 above)
Comment (swilly22) [reply src/execution_plan/ops/op_index_scan.c]: 👍

File: src/execution_plan/ops/op_delete.c (hunk 6 above)
Comment (swilly22) [reply src/execution_plan/ops/op_delete.c]: checking for the entity type?

File: src/execution_plan/ops/op_create.c (hunk 7 above)
Comment (swilly22) [reply src/execution_plan/ops/op_create.c]: I think so, currently only cartesian product issue a restart, which propagates upwards and so reset will never reach OpCreate

File: src/execution_plan/ops/op_create.c (hunk 8 above)
Comment (swilly22) [reply src/execution_plan/ops/op_create.c]: It might be that we can, I'll have to double check.

File: src/execution_plan/ops/op_conditional_traverse.c (hunk 9 above)
Comment (swilly22) [reply src/execution_plan/ops/op_conditional_traverse.c]: Impossible, if we're here then we've found a dest node from src node, and that's only possible if there exists an edge connecting the two, see comment at line 127,
additional calls to consume will be fulfilled by line 104 until all edges connecting src to dest are consumed, in which case a new set of edges is computed if such exists

File: src/execution_plan/ops/op_cond_var_len_traverse.c (hunk 10 above)
Comment (swilly22) [reply src/execution_plan/ops/op_cond_var_len_traverse.c]: 👍

File: src/arithmetic/arithmetic_expression.h (hunk 11 above)
Comment (swilly22) [reply src/arithmetic/arithmetic_expression.h]: 👍

File: src/execution_plan/ops/op_conditional_traverse.c (hunk 9 above)
Comment (jeffreylovitz) [reply src/execution_plan/ops/op_conditional_traverse.c]: :+1:

File: src/execution_plan/ops/op_create.c (hunk 7 above)
Comment (jeffreylovitz) [reply src/execution_plan/ops/op_create.c]: :+1:

File: src/execution_plan/ops/op_create.c (hunk 8 above)
Comment (swilly22) [reply src/execution_plan/ops/op_create.c]: Reason why we use a pointer to the newly created node, is because of the way edges are created,
newly created edges will retrieve their src and dest nodes from a record, they must point to the original nodes.
